modules/ml/src/main/java/org/apache/ignite/ml/recommendation/util/RecommendationDatasetData.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.recommendation.util;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.recommendation.ObjectSubjectRatingTriplet;
+import org.apache.ignite.ml.recommendation.RecommendationTrainer;
+import org.apache.ignite.ml.util.Utils;
+
+/**
+ * A partition {@code data} of a dataset required in {@link RecommendationTrainer}.
+ *
+ * @param <O> Type of an object.
+ * @param <S> Type of a subject.
+ */
+public class RecommendationDatasetData<O extends Serializable, S extends Serializable> implements AutoCloseable {
+    /** All ratings related to the partition. */
+    private final List<? extends ObjectSubjectRatingTriplet<O, S>> ratings;
+
+    /**
+     * Constructs a new instance of recommendation dataset data.
+     *
+     * @param ratings All ratings related to the partition.
+     */
+    public RecommendationDatasetData(List<? extends ObjectSubjectRatingTriplet<O, S>> ratings) {
+        this.ratings = Collections.unmodifiableList(ratings);
+    }
+
+    /**
+     * Calculates gradient of the loss function of recommendation system SGD training.
```

Comment:
I think it will be more convenient if you write gradient formulas here in javadoc explicitly. They are unreadable in code.

Reply:
I haven't found a convenient way to add a formula into Javadoc, but I can add a link to an article about it. Do you think it makes sense?

Reply:
Yes, it will be great!